File: e2e/go/engine_test.go
// E2e tests for category: engine
package e2e_test

import (
	"strings"
	"testing"

	pkg "github.com/kreuzberg-dev/kreuzcrawl"
)

func Test_EngineBatchBasic(t *testing.T) {
	// CrawlEngine with defaults batch scrapes like the free function
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'batch.completed_count' not available on result type
	// skipped: field 'batch.total_count' not available on result type
}

func Test_EngineCrawlBasic(t *testing.T) {
	// CrawlEngine with defaults crawls multiple pages like the free function
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'crawl.pages_crawled' not available on result type
	// skipped: field 'crawl.min_pages' not available on result type
}

func Test_EngineMapBasic(t *testing.T) {
	// CrawlEngine with defaults discovers URLs like the free function
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'map.min_urls' not available on result type
}

func Test_EngineScrapeBasic(t *testing.T) {
	// CrawlEngine with defaults scrapes a page identically to the free function
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	var metadataTitle string
	if result.Metadata.Title != nil {
		metadataTitle = *result.Metadata.Title
	}
	var metadataDescription string
	if result.Metadata.Description != nil {
		metadataDescription = *result.Metadata.Description
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %v", result.StatusCode)
	}
	if strings.TrimSpace(result.ContentType) != `text/html` {
		t.Errorf("equals mismatch: got %v", result.ContentType)
	}
	if strings.TrimSpace(metadataTitle) != `Engine Test` {
		t.Errorf("equals mismatch: got %v", metadataTitle)
	}
	if !strings.Contains(string(metadataDescription), `Testing the engine`) {
		t.Errorf("expected to contain %s, got %v", `Testing the engine`, metadataDescription)
	}
	if len(result.Links) < 1 {
		t.Errorf("expected >= 1, got %v", len(result.Links))
	}
	if result.Metadata.Headings != nil {
		if result.Metadata.Headings != nil {
			if len(*result.Metadata.Headings) < 1 {
				t.Errorf("expected >= 1, got %v", len(*result.Metadata.Headings))
			}
		}
	}
}

func Test_EngineStreamBasic(t *testing.T) {
	// CrawlEngine with defaults streams events like the free function
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'stream.has_page_event' not available on result type
	// skipped: field 'stream.has_complete_event' not available on result type
	// skipped: field 'stream.event_count_min' not available on result type
}

File: e2e/java/src/test/java/dev/kreuzberg/e2e/EngineTest.java
package dev.kreuzberg.e2e;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

/** E2e tests for category: engine. */
class EngineTest {
    @Test
    void testEngineBatchBasic() throws Exception {
        // CrawlEngine with defaults batch scrapes like the free function
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(2, result.batch().completedCount());
        assertEquals(2, result.batch().totalCount());
    }

    @Test
    void testEngineCrawlBasic() throws Exception {
        // CrawlEngine with defaults crawls multiple pages like the free function
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(3, result.crawl().pagesCrawled());
        assertTrue(result.crawl().minPages() >= 3, "expected >= 3");
    }

    @Test
    void testEngineMapBasic() throws Exception {
        // CrawlEngine with defaults discovers URLs like the free function
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertTrue(result.map().minUrls() >= 2, "expected >= 2");
    }

    @Test
    void testEngineScrapeBasic() throws Exception {
        // CrawlEngine with defaults scrapes a page identically to the free function
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(200, result.statusCode());
        assertEquals("text/html", result.contentType());
        assertEquals("Engine Test", result.metadata().title().orElse(""));
        assertTrue(result.metadata().descriptionContains().contains("Testing the engine"), "expected to contain: " + "Testing the engine");
        assertTrue(result.links().minCount() >= 1, "expected >= 1");
        assertEquals(1, result.headings().h1Count());
        assertEquals("Hello Engine", result.headings().h1Text());
    }

    @Test
    void testEngineStreamBasic() throws Exception {
        // CrawlEngine with defaults streams events like the free function
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(true, result.stream().hasPageEvent());
        assertEquals(true, result.stream().hasCompleteEvent());
        assertTrue(result.stream().eventCountMin() >= 3, "expected >= 3");
    }

}
